### PR TITLE
Lb index

### LINKED ISSUE #2300: py; lbgpu doesn't complain when accessing non-existing nodes
lbf[x,y,z].velocity
when  on of x,y,z > grid dim

### PR DESCRIPTION
**Pull request by KaiSzuttor**
Fixes #2300

Description of changes:
 - added a function to check validity of node index for LB
 - added testcase on interface level

PR Checklist
------------
 - [x] Tests?
   - [x] Interface
   - [ ] Core

### CODE REVIEW

File: src/core/Vector.hpp [hunk 1]
```diff
@@ -164,6 +164,7 @@ template <size_t n, typename Scalar> class Vector {
 // Useful typedefs
 
 typedef Vector<3, double> Vector3d;
+typedef Vector<3, int> Vector3dInt;
```
Comment (fweik) [inline src/core/Vector.hpp]: That should be called `Vector3i`. Please also use `using` over `typedef`.

**Comment by RudolfWeeber** [comment]
I thought the d in vector3d stood for double? then it would be vector3i.

Also, for the bugfix release, this pr should be split into a part that fixes the bug and will go into 4.0.1, and other changes, that won't.

File: src/core/grid_based_algorithms/lb.cpp [hunk 2]
```diff
@@ -1556,7 +1576,9 @@ int lb_lbnode_set_pop(int *ind, double *p_pop) {
   return 0;
 }
 
-int lb_lbnode_set_extforce_density(int *ind, double *f) { return -100; }
+int lb_lbnode_set_extforce_density(const Vector3dInt &ind, double *f) {
+  return -100;
```
Comment (fweik) [inline src/core/grid_based_algorithms/lb.cpp]: What does `-100` mean?

File: src/python/espressomd/utils.pxd [hunk 3]
```diff
@@ -92,6 +92,10 @@ cdef extern from "<limits>" namespace "std" nogil:
         T max()
 
 cdef extern from "Vector.hpp":
+    cppclass Vector3dInt:
```
Comment (fweik) [inline src/python/espressomd/utils.pxd]: Cython has in principal template support, (with the `[]` syntax, like `vector[double]`). Can you please try if you can get this to work without the code duplication with the `Vector3d`?

File: src/core/grid_based_algorithms/lb.cpp [hunk 4]
```diff
@@ -1167,7 +1167,25 @@ int lb_lbfluid_load_checkpoint(char *filename, int binary) {
   return ES_OK;
 }
 
-int lb_lbnode_get_rho(int *ind, double *p_rho) {
+bool lb_lbnode_is_index_valid(const Vector3dInt &ind) {
+  auto within_bounds = [](const Vector3dInt &ind, const Vector3dInt &limits) {
+    return ind < limits;
```
Comment (fweik) [inline src/core/grid_based_algorithms/lb.cpp]: I think this should be `ind >= {} && ind < limits`, as the indices are signed.

**Review by fweik — CHANGES_REQUESTED** [review]

File: src/core/Vector.hpp (hunk 1 above)
Comment (KaiSzuttor) [reply src/core/Vector.hpp]: ```using``` and ```typedef``` are equivalent and i adopted to the existing usage.

File: src/core/grid_based_algorithms/lb.cpp (hunk 2 above)
Comment (KaiSzuttor) [reply src/core/grid_based_algorithms/lb.cpp]: Who knows?

File: src/python/espressomd/utils.pxd (hunk 3 above)
Comment (KaiSzuttor) [reply src/python/espressomd/utils.pxd]: It's not possible, cython does not support non-type template parameters.

File: src/core/grid_based_algorithms/lb.cpp (hunk 2 above)
Comment (fweik) [reply src/core/grid_based_algorithms/lb.cpp]: I'm not an expert in these matters, but this does not look like a sensible function to me.

**Comment by KaiSzuttor** [comment]
@RudolfWeeber  what part should not go into bugfix release?

**Comment by RudolfWeeber** [comment]
On Tue, Oct 23, 2018 at 06:37:27AM -0700, Kai Szuttor wrote:
> @RudolfWeeber  what part should not go into bugfix release?
Everything which is not directly related to checkint ghe indices in LB.

To make sure that no new issues are introduced and to still be able to cherry-pick bugfixes with minimal edits, those fixes have to be as small as possible.
You can simply rebase your changes into two separate commits, one which introduces Vector3i and sorts out the lb index checking, and another with everything else.

File: src/python/espressomd/utils.pxd (hunk 3 above)
Comment (fweik) [reply src/python/espressomd/utils.pxd]:
```c++
template<typename T>
using Vector3 = Vector<3, T>;
```?

**Review by fweik — APPROVED** [review]

File: src/python/espressomd/utils.pxd (hunk 3 above)
Comment (KaiSzuttor) [reply src/python/espressomd/utils.pxd]: Oh, I misunderstood... that should be possible

**Comment by KaiSzuttor** [comment]
@RudolfWeeber I'm not sure how to go about this

**Comment by fweik** [comment]
@KaiSzuttor you can only import `from .lb cimport lb_lbnode_is_index_valid` if `LB`.

**Comment by RudolfWeeber** [comment]
On Tue, Oct 23, 2018 at 07:21:50AM -0700, Kai Szuttor wrote:
> @RudolfWeeber I'm not sure how to go about this
One way is to save the diff for the entire pr, apply it to a fresh
branch forked from upstream.
Then you can add use 
git add
to add only the part related to lb (Vecotr.hpp, lb*.?pp, lb*.cu,
src/python/espressomd/lb.p?)
and commit that.
This can then be cherry-picked into the bugfix branch.
In a 2nd commit, you can add the use of the Vector3i in all the
other places.

**Comment by KaiSzuttor** [comment]
@RudolfWeeber I changed the commit history accordingly (git rebase espressomd/python; git reset upstream/python, then recommit).

**Pull request merged**